annotations_ingester/annotations_generator.py

```diff
@@ -0,0 +1,196 @@
+import logging
+import re
+import sys
+from pathlib import Path
+from typing import Sequence
+
+import boto3
+import botocore
+import pystac
+from pulsar import Message
+from pystac import Catalog, Collection, STACTypeError
+from rdflib import DCAT, DCTERMS, RDF, Graph, Literal, URIRef
+
+from annotations_ingester.messagers import CatalogueSTACChangeMessager, Messager
+
+from rdflib import Graph
+
+
+CATALOGUE_PUBLIC_BUCKET_PREFIX = "/catalogs/supported-datasets/ceda-stac-catalogue/collections/"
+
+
+def is_file_immutable(file_contents):
+    lowercase = file_contents.lower()
+    if 'uuid' in lowercase and re.search('[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}', lowercase):
+        logging.info("File is immutable")
+        return True
+
+    logging.info("File is not immutable")
+    return False
+
+
+
+def convert_format(file_path: str, format: str):
+    g = Graph()
+    g.parse(file_path)
+    return g.serialize(format=format)
+
+
+def download_s3_file(file_name: str, bucket: str, s3_client):#, file_path: str):
+    s3_client.download_file(bucket, file_name, file_name)
+    file_contents = open(file_name).read()
+
+    return file_contents
+
+    # output_file = Path(file_path)
+    # output_file.parent.mkdir(exist_ok=True, parents=True)
+    # with open(file_path, 'w') as f:
+    #     f.write(file_contents)
+
+
+class AnnotationsMessager(CatalogueSTACChangeMessager):
+    """
+    Generates basic DCAT for catalogue entries. Supports Catalogs and Collections and is
+    intended only to be sufficient for finding QA information linked to a dataset.
+
+    The output is sent to the cataloge public static files bucket under a path matching the
+    catalogue API endpoint for the dataset. For example:
+        /catalogue/
+    """
+
+    def __init__(self, catalogue_public_bucket: str):
+        self._dest_bucket = catalogue_public_bucket
+
+    def process_delete(self):
+        pass
+
+    def process_stac_update(
+        self,
+        cat_path: str,
+        stac: dict,
+        **kwargs,
+    ) -> Sequence[Messager.Action]:
+        pass
+
+
+
+    def process_msg(self, msg: Message) -> Sequence[Messager.Action]:
```

Comment:
Instead of overriding this (`process_msg` is already implement in CatalogueChangeMessager from which CatalogueSTACChangeMessager inherits) you should override `process_update_stac`. By doing that, most of the lines of this method aren't required any more: parsing the message, looping over keys, downloading files and handling exceptions are all done by the superclasses.

Reply:
Updated